File: apps/web/src/lib/imageModels.ts
import type { ImageModel, RenderQualityChoice } from "./types";

export const DEFAULT_IMAGE_MODEL: ImageModel = "gpt-image-2";
export const IMAGE_MODEL_OPTIONS: ReadonlyArray<{ value: ImageModel; label: string }> = [
  { value: "gpt-image-2", label: "GPT Image 2" },
  { value: "gpt-image-2.5-flare", label: "GPT Image 2.5 Flare" },
  { value: "gpt-image-2.5-sunburst", label: "GPT Image 2.5 Sunburst" },
];

const QUALITY_OPTIONS: ReadonlyArray<{ value: RenderQualityChoice; label: string }> = [
  { value: "low", label: "Low" },
  { value: "medium", label: "Medium" },
  { value: "high", label: "High" },
  { value: "xhigh", label: "Xhigh" },
  { value: "max", label: "Max" },
];

export function normalizeImageModel(value: unknown): ImageModel {
  return IMAGE_MODEL_OPTIONS.find((option) => option.value === value)?.value ?? DEFAULT_IMAGE_MODEL;
}

export function imageQualityOptions(model: unknown) {
  return normalizeImageModel(model) === DEFAULT_IMAGE_MODEL
    ? QUALITY_OPTIONS.slice(0, 3)
    : QUALITY_OPTIONS;
}

export function normalizeImageQuality(value: unknown, model: unknown): RenderQualityChoice {
  return imageQualityOptions(model).find((option) => option.value === value)?.value ?? "high";
}

export function imageParamsForReroll(generation: {
  requested_params?: Record<string, unknown> | null;
  request_params?: Record<string, unknown> | null;
  effective_params?: Record<string, unknown> | null;
}) {
  const params = generation.requested_params ?? generation.request_params ?? generation.effective_params ?? {};
  const model = normalizeImageModel(params.model ?? params.image_model);
  return { model, render_quality: normalizeImageQuality(params.render_quality, model) };
}

File: packages/core/lumen_core/image_models.py
"""Supported Image API models and quality capabilities."""

from typing import Literal, get_args

ImageModel = Literal["gpt-image-2", "gpt-image-2.5-flare", "gpt-image-2.5-sunburst"]
ImageRenderQuality = Literal["auto", "low", "medium", "high", "xhigh", "max"]
DEFAULT_IMAGE_MODEL: ImageModel = "gpt-image-2"
MAX_IMAGE_COUNT = 10
IMAGE_MODELS = frozenset(get_args(ImageModel))
IMAGE_RENDER_QUALITIES = frozenset(get_args(ImageRenderQuality))


def validate_image_model_quality(model: str, quality: str) -> None:
    if model not in IMAGE_MODELS:
        raise ValueError(f"unsupported image model: {model}")
    if quality not in IMAGE_RENDER_QUALITIES:
        raise ValueError(f"unsupported image quality: {quality}")
    if model == DEFAULT_IMAGE_MODEL and quality in {"xhigh", "max"}:
        raise ValueError(f"{model} does not support {quality} quality")
